Add `-Xlint:pattern-shadow` to lint pattern varids which are backquotable

som-snytt opened the pull request:
Lint `-Xlint:pattern-shadow` warns when a pattern has a variable that may have been intended to match a value in scope using "backtick" notation.
```
def compare(x: Any, y: Any) = x match { case y => true }  // intended case `y` =>
```
For the common idiom of "refining" a value through pattern matching, there is no warning if the value in scope is used in the selector expression:
```
def asString(x: Any) = x match { case x: String => x case _ => stringOf(x) }

def f(x: X) = x.alias match { case x => } // does not check that x in pattern conforms to the input parameter

def f(x: X) = g(x) match { case x => } // arbitrary functions in x may extract arbitrary values without incurring a warning
```
This exception is permissive. For example, tuple arguments need not "align" in the pattern:
```
def swap(x: Any, y: Any) = (x, y) match { case (y, x) => }
```

Fixes scala/bug#11850

som-snytt commented:
Addresses https://github.com/scalapuzzlers/scalapuzzlers.github.com/pull/154

som-snytt commented:
Also patdefs, cf https://github.com/scala/bug/issues/7530

som-snytt commented:
Clunky lint would be nicer as lightweight 3rd party linter and lint suppression via `-Wconf` and `@nowarn`.

som-snytt commented:
Was I supposed to revisit this? Summer '20 was too early in pandemic to be morbidly depressed. My system of adding annotations in the PR title must have broken down.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4628,6 +4628,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             tree setSymbol sym setType sym.tpeHK
 
           case Bind(name: TermName, body)  =>
+            if (settings.warnPatternShadow && !tree.hasAttachment[NoWarnAttachment.type] && context.isNameInScope(name))
```
Why the `NoWarn` filter?

som-snytt replied:
All questions will be answered after I finish linting "literal boolean args should be named". That will be after I have coffee.

som-snytt replied:
The attachment represents user code
```
case x @ _ =>
```
for which we don't want to propose
```
case `x` @ _ =>
```
Backquotes in that case is useful for
```
case `type` @ _ =>
```

lrytz reviewed (commented):
This looks useful - did you try it on the compiler / library to see how often it fires?

som-snytt commented:
I have a second branch where I may have fixed some project source, because this feels familiar:
```
    def mkAnnotation(tree: Tree): Tree = tree match {
      case SyntacticNew(Nil, SyntacticApplied(SyntacticAppliedType(_, _), _) :: Nil, noSelfType, Nil) =>
        tree
```
where `noSelfType` is not
```
Name noSelfType is already introduced in an enclosing scope as object noSelfType in trait Trees. Did you intend to match it using backquoted `noSelfType`?
```
Why doesn't this come up as an obvious reason not to name all values with lower case?

The other interesting aspect is that the var is unused on the RHS. I may have already considered that as a criterion: surely if I use the var, I intended to use it? unless I unintended the other element.

som-snytt commented:
Forgot to restarr locally before pushing, d'oh. This day is already much too long, there will be more days. I almost wrote morose days. I noticed this PR began on my birthday, two days before lockdown. An age ago, so to speak.

som-snytt commented on src/compiler/scala/reflect/reify/phases/Metalevels.scala:
```diff
@@ -128,13 +128,13 @@ trait Metalevels {
     //    to the contrast, reified types (i.e. synthetic typetags materialized by Implicits.scala) always stay on the same metalevel as their enclosing code
     override def transform(tree: Tree): Tree = tree match {
-      case TreeSplice(ReifiedTree(universe, mirror, symtab, rtree, tpe, rtpe, concrete)) =>
+      case TreeSplice(ReifiedTree(_universe, _mirror, symtab, rtree, tpe, rtpe, _concrete)) =>
```
example giving up

som-snytt commented on src/compiler/scala/tools/nsc/ast/DocComments.scala:
```diff
@@ -458,7 +458,7 @@ trait DocComments { self: Global =>
         def findIn(sites: List[Symbol]): Type = sites match {
           case List() => NoType
-          case site :: sites1 => select(site.thisType, name, findIn(sites1))
+          case h :: sites1 => select(h.thisType, name, findIn(sites1))
```
could be `h :: t`, as `sites1` is not descriptive, or `rest`.

som-snytt commented on src/compiler/scala/tools/nsc/ast/DocComments.scala:
```diff
@@ -543,8 +543,8 @@ trait DocComments { self: Global =>
-          case tp1 @ TypeRef(pre, sym, args) if (sym.name.length > 1 && sym.name.startChar == '$') =>
-            subst(sym, aliases, aliasExpansions) match {
+          case tp1 @ TypeRef(_, sym0, args) if (sym0.name.length > 1 && sym0.name.startChar == '$') =>
+            subst(sym0, aliases, aliasExpansions) match {
```
note the following `sym1`; however, not very satisfactory. But the difficulty is everywhere in a generic model with trees, symbols, args, types. Occasionally, a domain-specific name is obvious.

som-snytt commented on src/compiler/scala/tools/nsc/ast/Trees.scala:
```diff
@@ -196,9 +196,9 @@ trait Trees extends scala.reflect.internal.Trees { self: Global =>
+    case SelectFromArray(qualifier, selector, erasure0) =>
       transformer.treeCopy.SelectFromArray(
-        tree, transformer.transform(qualifier), selector, erasure)
+        tree, transformer.transform(qualifier), selector, erasure0)
```
because of equals semantics, hard to say they couldn't possibly intend erasure phase

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -3395,9 +3395,10 @@ self =>
-      case result       => result
+    def templateBody(isPre: Boolean) = {
+      val selfBody = inBraces(templateStatSeq(isPre = isPre))
+      if (selfBody._2.isEmpty) selfBody.copy(_2 = EmptyTree.asList)
```
there's a reason no one does that. Prefer to just rename self.

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala:
```diff
@@ -91,10 +91,10 @@ abstract class TreeBuilder {
     def alternatives(t: Tree): List[Tree] = t match {
-      case Alternative(ts)  => ts
-      case _                => List(t)
+      case Alternative(alts) => alts
```
sample useful clarification (domain-specific)

som-snytt commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -684,15 +684,15 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
                 accept('>')
                 assert(xs.length > 0, tp)
-                debuglogResult("new existential")(newExistentialType(existentials.toList, typeRef(pre, classSym, xs.toList)))
+                debuglogResult("new existential")(newExistentialType(existentials.toList, typeRef(pre, tsym, xs.toList)))
```
maybe the intervening 30 LOC could be moved to a method, then these refs are closer to the pattern they come from

som-snytt commented on src/compiler/scala/tools/nsc/tasty/bridge/TypeOps.scala:
```diff
@@ -418,7 +418,7 @@ trait TypeOps { self: TastyUniverse =>
           def elemType(tp: Type): Type = tp.dealiasWiden match {
             // case tp: AndOrType => tp.derivedAndOrType(elemType(tp.tp1), elemType(tp.tp2))
-            case tp: u.RefinedType => u.intersectionType(tp.parents.map(elemType))
+            case tdw: u.RefinedType => u.intersectionType(tdw.parents.map(elemType))
```
sample where `q.m match { case q: Q => }` could be excluded, like `x match { case x: X => }`.

som-snytt commented on src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:
```diff
@@ -1744,9 +1744,9 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
           }
 
-        case Template(parents, self, body) =>
+        case Template(parents, self, tbody) =>
```
if less lazy, I would have renamed the ambient `body` element to something more descriptively specific. That allows inner names to be generic, such as body here.

som-snytt commented on src/compiler/scala/tools/nsc/transform/patmat/PatternMatching.scala:
```diff
@@ -66,11 +66,11 @@ trait PatternMatching extends Transform
 
         // setType origTp intended for CPS -- TODO: is it necessary?
-        val translated = translator(sel.pos).translateMatch(treeCopy.Match(tree, transform(sel), transformCaseDefs(cases)))
+        val translated = translator(selector.pos).translateMatch(treeCopy.Match(tree, transform(selector), transformCaseDefs(cases)))
```
only a relic from touching the code here. The check wound up in refchecks.

som-snytt commented on src/compiler/scala/tools/nsc/transform/patmat/Solving.scala:
```diff
@@ -308,7 +308,7 @@ trait Solving extends Logic {
     }
 
-    def eqFreePropToSolvable(p: Prop): Solvable = {
+    def eqFreePropToSolvable(prop: Prop): Solvable = {
```
same idea of longer names outside, shorter names inside.

som-snytt commented on src/interactive/scala/tools/nsc/interactive/Pickler.scala:
```diff
@@ -106,32 +106,32 @@ object Pickler {
-      case UnpickleSuccess(x) => UnpickleSuccess(f(x))
-      case f: UnpickleFailure => f
+      case UnpickleSuccess(x)    => UnpickleSuccess(f(x))
+      case fail: UnpickleFailure => fail
```
💯

som-snytt commented on src/library/scala/collection/immutable/IntMap.scala:
```diff
@@ -32,9 +32,9 @@ private[immutable] object IntMapUtils extends BitOperations.Int {
-    case (left, right) => IntMap.Bin(prefix, mask, left, right)
+    case (_, IntMap.Nil) => left
+    case (IntMap.Nil, _) => right
+    case _ => IntMap.Bin(prefix, mask, left, right)
```
the previous no longer warns, but maybe it is an improvement

som-snytt commented on src/library/scala/collection/immutable/IntMap.scala:
```diff
@@ -409,8 +409,8 @@ sealed abstract class IntMap[+T] extends AbstractMap[Int, T]
-  def unionWith[S >: T](that: IntMap[S], f: (Int, S, S) => S): IntMap[S] = (this, that) match{
-    case (IntMap.Bin(p1, m1, l1, r1), that@(IntMap.Bin(p2, m2, l2, r2))) =>
+  def unionWith[S >: T](that: IntMap[S], f: (Int, S, S) => S): IntMap[S] = (this, that) match {
+    case (IntMap.Bin(p1, m1, l1, r1), IntMap.Bin(p2, m2, l2, r2)) =>
```
presumably no longer would warn, but is one form preferable? it compiles. Also I presume `that` was a `Bin` before, but the syntax is noisy.

som-snytt commented on src/library/scala/collection/mutable/HashMap.scala:
```diff
@@ -461,7 +461,7 @@ class HashMap[K, V](initialCapacity: Int, loadFactor: Double)
       val nd = table(idx) match {
         case null => null
-        case nd => nd.findNode(key, hash)
+        case nd0  => nd0.findNode(key, hash)
```
should not warn

som-snytt commented on src/library/scala/collection/mutable/LinkedHashMap.scala:
```diff
@@ -135,7 +135,7 @@ class LinkedHashMap[K, V]
       val nd = table(idx) match {
         case null => null
-        case nd => nd.findEntry(key, hash)
+        case nd0  => nd0.findEntry(key, hash)
```
ditto

som-snytt commented on src/reflect/scala/reflect/internal/AnnotationInfos.scala:
```diff
@@ -358,26 +358,28 @@ trait AnnotationInfos extends api.Annotations { self: SymbolTable =>
+        else throw new Exception(s"unexpected annotation type $atp: only subclasses of StaticAnnotation and ClassfileAnnotation are supported")
+      case _ =>
+        throw new Exception("""unexpected tree shape: only q"new $annType(..$args)" is supported""")
+    }
```
example denest to avoid shadow

som-snytt commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -743,7 +743,7 @@ trait Printers extends api.Printers { self: SymbolTable =>
           print("@")
           core match {
-            case Select(New(tree), _) => print(tree)
+            case Select(New(corer), _) => print(corer)
```
is that like `apple corer` or like `more core than`?

som-snytt commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -23,7 +23,7 @@ import java.lang.System.{lineSeparator => EOL}
 import scala.annotation.{nowarn, tailrec}
 
-trait Printers extends api.Printers { self: SymbolTable =>
+trait Printers extends api.Printers { symbolTable: SymbolTable =>
```
ok that works. It's used 1200 LOC later.

som-snytt commented on src/reflect/scala/reflect/internal/ReificationSupport.scala:
```diff
@@ -114,7 +114,7 @@ trait ReificationSupport { self: SymbolTable =>
 
     def mkAnnotation(tree: Tree): Tree = tree match {
-      case SyntacticNew(Nil, SyntacticApplied(SyntacticAppliedType(_, _), _) :: Nil, noSelfType, Nil) =>
+      case SyntacticNew(Nil, SyntacticApplied(SyntacticAppliedType(_, _), _) :: Nil, `noSelfType`, Nil) =>
```
old bug alert, presumably, but maybe the name was merely documentary.

som-snytt commented on src/reflect/scala/reflect/internal/pickling/UnPickler.scala:
```diff
@@ -284,12 +284,12 @@ abstract class UnPickler {
+        case i =>
           privateWithin = NoSymbol
-          inforef = index
+          inforef = i
```
small names for small scopes

som-snytt replied:
actually this is just initializer for the vars.

som-snytt commented on src/reflect/scala/reflect/internal/tpe/TypeComparers.scala:
```diff
@@ -413,11 +413,11 @@ trait TypeComparers {
-      case (tp1, et2: ExistentialType)                                      => et2.withTypeVars(isSubType(tp1, _, depth), depth)
-      case (et1: ExistentialType, tp2)                                      =>
+      case (ntp1, et2: ExistentialType)                                     => et2.withTypeVars(isSubType(ntp1, _, depth), depth)
+      case (et1: ExistentialType, ntp2)                                      =>
```
ok but alignment, don't mess up their stuff

som-snytt commented on src/repl-frontend/scala/tools/nsc/interpreter/shell/LoopCommands.scala:
```diff
@@ -144,7 +144,7 @@ trait LoopCommands {
                 CompletionResult(buffer, cursor = 1, List(CompletionCandidate(completion)), "", "")
             }
-          case cmd :: rest =>
+          case cmd :: _ =>
```
if you don't use it, lose it

som-snytt commented:
Maybe this PR does too much work. I'll reassess locally.

som-snytt commented:
> However, I don't have an offhand idea why my PR fails the build.

Just because Seth marked it draft doesn't mean I won't take suggestions as to what the hell is going on. Maybe we need a label for that.

som-snytt commented:
ok it is my fault
```
[info] Test scala.tools.nsc.backend.jvm.OptimizedBytecodeTest.t2171 started

  | => junit / Test / testOnly 12418s
```

som-snytt commented:
After splitting the commit, I'm still testing just the "easy edits" and one of them broke the optimizer test!

Why do we say that "naming is hard" instead of "naming is evil". If Adam had not gone around naming things, do you think he would have eaten the apple? Instead of saying jocularly, "An apple a day...!", he would have scrutinized the type of the fruit and recognized its true nature.

lrytz commented:
> one of them broke the optimizer test

I can take a look if that helps, you could push it on a branch.

There might still be cases of cross-talk between optimizer tests, we reuse a Global instance. I fixed some issues, but possibly not all.

som-snytt commented on src/library/scala/collection/concurrent/TrieMap.scala:
```diff
@@ -343,19 +343,18 @@ private[collection] final class INode[K, V](bn: MainNode[K, V], g: Gen, equiv: E
                       case tn: TNode[K, V] @uc =>
-                        val ncn = cn.updatedAt(pos, tn.copyUntombed, gen).toContracted(lev - 5)
+                        val ncn = pm.updatedAt(pos, tn.copyUntombed, gen).toContracted(lev - 5)
                         if (!parent.GCAS(cn, ncn, ct))
```
OK I found the missing rename. UGH! A word I don't employ lightly.

som-snytt commented:
Thanks @lrytz it was just a missing rename. I split the commit and bisected.

My feature proposal for an IDE is to tell me when a referent changes between runs.

"can't hardly wait" (1998) was playing during the bisect. Many '90s actors! Jason Segel is "Watermelon guy"! So on that level, it was totally worth the effort.

I'm sorry I didn't save someone's visual pun on social media, sigma notation for "summing up the '90s".

som-snytt commented:
The error says some edits are missing? I `git stash pop`ped everything! I'll try to amend it during the next episode of Dawson's Creek. Pacey and Joey are going to tell Dawson!

`git stash poop`

som-snytt commented:
Wait, that's it? I thought I had more previous edits to apply, but I'll take it. @lrytz

lrytz commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -262,11 +262,13 @@ self =>
+          in.healBraces() match {
+            case Nil => showSyntaxErrors(); firstErr
+            case ps  => withPatches(ps).parse()
+          }
```
I assume the warning fired on `case patches` and the rest of the refactoring is scouting?

som-snytt replied:
Yes. It borders on "abuse of match", but I like that the names are meaningful labels for the branches. I see `ps` is not great, although the scopes are small. Maybe I will revert.

lrytz commented on src/compiler/scala/reflect/reify/phases/Metalevels.scala:
```diff
@@ -16,7 +16,7 @@ package phases
 
 trait Metalevels {
-  self: Reifier =>
+  this: Reifier =>
```
Related to the new warning?

som-snytt replied:
Not here but elsewhere it clashes with the `self` in `Template`. Why introduce a name unnecessarily? I wonder if `-Wunused` flags it? :bulb:

lrytz commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -648,13 +648,11 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
-              typeRef(processInner(pre.widen), sym, args)
-            case _ =>
-              tp
+            case TypeRef(_pre, _sym, _args) if !_sym.isStatic => typeRef(processInner(_pre.widen), _sym, _args)
```
Hmm... Seeing diffs like this makes me wonder. Naming things is already absorbing a lot of time when coding and the new lint will intensify that. People will then resort to fixes like this one, making the code look worse.

I think this is a good example: writing `case TypeRef(pre, sym, args)` is very common in this codebase and everybody expects it to bind new names. If we can no longer write this it's distracting.

Here's an example which you pointed out as a good one (https://github.com/scala/scala/pull/8806/files#r1428583071):

<img width="568" alt="image" src="https://github.com/scala/scala/assets/119636/b20bbe68-e7b0-4652-b57d-ac92fe013a34">

I agree this is a good change.

Unfortunately I don't have an idea right now to tell the good and worse cases apart...

som-snytt replied:
💡 ⚡ it occurs to me that you must allow canonical names as determined by constructor pattern (case params) or extractor (unapply params).  Then we're not just looking at the binding but the context. I'll give that a shot.

However, I did convince myself that the shadowing problem is real. What is `tree`? That's more urgent when `fun` gets recycled a few times. My provisional fix was to stop inventing names and use `_foo` to mean a kind of "pass-through", if it's used immediately in RHS expr and not referenced 100 LOC away. There was one `classSym` like that, not sure I addressed that particular one.

The motivating case is patvar shadowing other kinds of definitions in scope, so maybe that is another dimension.

Also on naming, it wasn't always immediately obvious what a "good name" would be without a bit of study (which I avoided for reasons of scope).

My erroneous refactor shows how treacherous it can be. (I did not use intellij to perform renames, so maybe it falls under "things you can't do without an IDE".)

Another open question is the form `f(x) match { case x => }`, where x is an input to the scrutinee.

lrytz reviewed (commented):
First look...

som-snytt commented on src/compiler/scala/reflect/reify/phases/Metalevels.scala:
```diff
@@ -128,13 +128,16 @@ trait Metalevels {
-        if (reifyDebug) println("entering inlineable splice: " + tree)
-        val inlinees = symtab.syms filter (_.isLocalToReifee)
-        inlinees foreach (inlinee => symtab.symAliases(inlinee) foreach (alias => inlineableBindings(alias) = symtab.symBinding(inlinee)))
+      case TreeSplice(ReifiedTree(_universe, _mirror, symtab, _rtree, _tpe, _rtpe, _concrete)) =>
```
This is where I decided to use underscore as special marker. Only `symtab` is "used" and replaced. Maybe the better idiom is `case rt @ ReifiedTree(_, ..., symtab, ...) => rt.copy(symtab = symtab1)`. I was deterred by `withinSplice` etc. I didn't want to change anything but syntax.

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -3410,8 +3412,8 @@ self =>
-      case (self, Nil)  => (self, EmptyTree.asList)
-      case result       => result
+      case (selfTypeVal, Nil) => (selfTypeVal, EmptyTree.asList)
+      case result => result
```
I think I also tried this as `pair.copy(_2 = EmptyTree.asList)`. I wonder if val looks better, `val pair @ (selfie, stats) = f()`

som-snytt commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -648,13 +648,11 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
           }
           def processClassType(tp: Type): Type = tp match {
-            case TypeRef(pre, classSym, args) =>
+            case TypeRef(pre, typerefSym, args) =>
```
this is where the usage is far away. So it's the combination of shadowing and distance. There ought to be a complexity measure that accounts for it.

som-snytt commented:
Dotty using underscore to mean "pass-through":
```
private class Printer(val seen: Seen, _ctx: Context) extends RefinedPrinter(_ctx):
```

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -264,9 +264,13 @@ self =>
+          showSyntaxErrors()
+          firstTry
+        }
+        else withPatches(patches).parse()
```
still not satisfying. Maybe `in.healBraces().pipe(patches => ...)`.

som-snytt replied:
the "one-liner" is compelling? `if (patches.isEmpty) firstTry.tap(_ => showSyntaxErrors())` is expressive

som-snytt commented on src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:
```diff
@@ -699,9 +697,13 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
+              case classSym =>
+                assert(!classSym.isOverloaded, classSym.alternatives)
+                classSym.tpe_*
+            }
```
oh, clever. The encapsulation permits the ident `classSym` to be used later (in a pattern). However, the idiom `case _: ObjectClass.type` looks strange FSR. I would prefer "abuse of match" formulation, `case classSym if classSym eq ObjectClass` or better yet: `val classTpe = { val classSym = ...; if - else }`

som-snytt commented on src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala:
```diff
@@ -648,7 +648,7 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
             def lengthMax3(l: List[List[TreeMaker]]): Int = l match {
               case a :: b :: c :: _ => 3
-              case cases            => cases.map {
+              case kases            => kases.map {
```
better to shadow at `def lengthMax3(cases: List[...])` which actually represents the shadowing, and the scrutinee exclusion falls out naturally

also `kases` was too `kute`

not sure if the pattern match is better than lengthCompare?

som-snytt commented on src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:
```diff
@@ -911,7 +907,7 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
     subclasses foreach {
       env =>
-      val spc      = specializedClass(env, decls1)
+      val spc      = toSpecializedClass(env, decls1)
```
avoiding name clash here

som-snytt commented:
Net about 70 LOC, plus 200 lines in test.

som-snytt commented:
This is the one where I said, Oh, I started it on my birthday two days before lockdown. I think it's ready for last chance review.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -2708,16 +2709,23 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-      def finish(cases: List[CaseDef], matchType: Type) =
-        treeCopy.Match(tree, selector1, cases) setType matchType
+      def finish(cases: List[CaseDef], matchType: Type) = {
+        if (!isPastTyper && settings.warnPatternShadow && !context.owner.isSynthetic && selector.symbol != null)
```
What is the `selector.symbol != null` check for?

som-snytt replied:
probably defensive coding. `42 match case i =>`.

som-snytt replied:
oh wait, this typer. It should set the shadow attachment; RefChecks ignores the selector if necessary. I'll see what turns up when I remove that condition.

som-snytt replied:
update: seems to be pattern-matching anonymous functions and selectors with an annotation.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -2048,6 +2048,67 @@ abstract class RefChecks extends Transform {
+                override def traverse(t: Tree): Unit = t match {
+                  case Apply(_, args) =>
+                    treeInfo.dissectApplied(t).core.tpe match {
+                      case MethodType(ps, _) =>
+                        foreach2(ps, args) { (p, arg) =>
+                          absolved ::= p.name
+                          try traverse(arg)
+                          finally absolved = absolved.tail
+                        }
```
This doesn't seem to work for extractors

```scala
// OK
scala> def f(x: Int) = "".toIntOption match { case Some(x) => 0; case _ => 1 }
                                                        ^
       warning: Name x is already introduced in an enclosing scope as value x. Did you intend to match it using backquoted `x`?

// OK
scala> def f(value: Int) = "".toIntOption match { case Some(value) => 0; case _ => 1 }

scala> object Xt { def unapply(spunk: Option[Int]) = spunk }

// OK
scala> def f(x: Int) = "".toIntOption match { case Xt(x) => 0; case _ => 1 }
                                                      ^
       warning: Name x is already introduced in an enclosing scope as value x. Did you intend to match it using backquoted `x`?

// Shbould not warn
scala> def f(spunk: Int) = "".toIntOption match { case Xt(spunk) => 0; case _ => 1 }
                                                          ^
       warning: Name spunk is already introduced in an enclosing scope as value spunk. Did you intend to match it using backquoted `spunk`?
```

som-snytt replied:
That was intentional because the unapply arg is not what is extracted.

lrytz replied:
Ah, yeah, of course 😵

lrytz reviewed (commented):
This a lot better now! I can imagine it being useful and not too distracting.

A bit of a concern is all the built in heuristics, it's going to be difficult for users to understand / reverse-engineer when shadowing is allowed. Maybe something high-level enough can be included in the lint help text?

As a general idea, what do you think about adding a tag to all lint messages, so that it's easy to tell if a warning is a lint, and from which category?

som-snytt commented:
This is a linty lint, as it pertains to a weak bit of syntax (the var id in a pattern) and is just a nag to say, did you mean...?

Hopefully, it also aids readability because it reduces ambiguity.

The two categories of "idioms deemed safe in this respect" are refining the scrutinee and constructor pattern args.

Examples of refining: `x match { case x: Int => }`, `(x, y) match { case (x: Int, y: Int) => }`.

Missing heuristic: `x.n match { case x => }`. Often `n` represents some facet of `x`, so it might be enough to say that if they are compatible types (conforming) then take this idiom as "refining x".

The goal of the PR is to determine whether it's effective enough to be worth porting, either to scalafix or scala 3.

I noticed that it's annoying to say "shadowed by value x", it should at least give a line number. An IDE could show both locations.

Per your suggestion, it sounds like we need `E101` numbers for errors and `L32` numbers for lints. It's tempting to backport the best of "awesome messages".

som-snytt committed:
Lint shadowing pattern variables

som-snytt commented on test/files/neg/annots-constant-neg.check:
```diff
@@ -45,7 +45,7 @@ Test.scala:43: error: annotation argument needs to be a constant; found: java.la
 Test.scala:48: error: nested classfile annotations must be defined in java; found: inline
   @JAnn(value = 0, c = Array(new inline)) def t18 = 0 // err
-                             ^
+   ^
```
this caret shift is wrong, and is due to an edit related to shadowing

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4086,8 +4094,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-        case ann @ Apply(Select(New(tpt), nme.CONSTRUCTOR), _) if isJava =>
-          val annInfo = typedAnnotation(ann, None, mode)
+        case Apply(Select(New(tpt), nme.CONSTRUCTOR), _) if isJava =>
+          val annInfo = typedAnnotation(tree, None, mode)
```
this edit misses a second usage a few lines later. The real danger is not shadowing but unshadowing.

lrytz replied:
Ups.. the end result is less ambiguity, but there's a risk that this happens to users while they address the new warnings.

som-snytt replied:
IDE could help with "atomic" refactor, but also incremental compilation could notice a change in attribution. (Previously, I've wanted incremental compilation to track something; I haven't kept a list.)

som-snytt committed:
Tweak library

som-snytt committed:
Tweak reflect

som-snytt commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -5345,23 +5345,23 @@ trait Types
   private[scala] final def typeIsNothing(tp: Type): Boolean =
     tp.dealias match {
-      case PolyType(_, tp) => typeIsNothing(tp)
+      case PolyType(_, resultType) => typeIsNothing(resultType)
```
these warnings go away by taking tp in the selection `tp.dealias` (and the types are OK). Very nice.

som-snytt commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -4372,19 +4372,19 @@ trait Types
     elementExtract(container, tp) match {
       case NoType => None
-      case tp => Some(tp)
+      case oktype => Some(oktype)
```
could allow `x.f(y, z) match { case x | y | z  => }` on the same criteria as `tp.dealias`. Not going to fix this now, maybe later. Oh but this code needs to pass the lint, so I guess I will the source for now.

som-snytt replied:
💡

som-snytt committed:
Tweak nsc

som-snytt committed:
Tweak more

som-snytt commented:
20 fewer files in the diff. Subproject tweaks are split out again.

som-snytt committed:
Tweak junit

som-snytt committed:
Handle typed selector

som-snytt commented:
Now I want it to see that `ValEq` is an extractor with an apply, and use param names from the apply, but I will not add that tonight.

som-snytt committed:
Partial Functions resisting

som-snytt committed:
Tweak junit

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -2048,6 +2048,73 @@ abstract class RefChecks extends Transform {
+                  for (cdef <- cases) check(cdef.pat, core.symbol)
+                case _ => checkAllCases()
+              }
+            checkShadowingPatternVars()
```
```suggestion
            if (settings.warnPatternShadow) checkShadowingPatternVars()
```

lrytz reviewed (commented):
> I noticed that it's annoying to say "shadowed by value x", it should at least give a line number

👍 and / or tell the owner even if `x` is a local.

som-snytt committed:
Add line number if not a member

som-snytt commented:
I think it would be nice to say `parameter x` when it is a parameter. The table in Symbols distinguishes the hard-core name for a thing from its pop name. But the breaking tests include ambiguous implicit, where you only care about the types, but to say `parameter foo and value bar` makes it look like "storage class" is significant. I will reconsider after looking at messages in Scala 3.

Otherwise, just add line number for locals and parameters. Hopefully, user can locate members.

Note that is not necessary to say `value x`, as we know it is a term (to be a candidate for the pattern) and we know its name, which was already stated. Finally, I restored the phrase because you get `object x` and also if you have debug turned up, you get `-uniqid`, which might be necessary to disambiguate some edge case.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -2048,6 +2048,78 @@ abstract class RefChecks extends Transform {
+                      case p => check(p, selectorSymbol)
+                    }
+                // or for `tp.dealiased match case tp =>` check `tp` if it looks like a derived value, but not `this.x` or `x.asInstanceOf`
+                case treeInfo.Applied(Select(core, _), _, _) if core.symbol != null && !core.symbol.isThisSym && selector.tpe <:< core.tpe.widen =>
```
What do you think about generalizing it to "shadowed symbol shows up in scrutinee"?

https://github.com/som-snytt/scala/compare/issue/11850...lrytz:scala:pr8806?expand=1

som-snytt replied:
I thought about it but was too lazy to experiment. I considered that `n + 1 match case n` should be the same as `1 + n`. Requiring a conforming type means it doesn't take arbitrary `f(x) match case x`. Generalizing is easier to explain and less fussy. OK, I will try it out.

som-snytt replied:
OK, I already forgot that I intended to follow up with an experiment.

som-snytt committed:
Loosey-goosey implication in selector

If the shadowed element appears anywhere in the selector expression,
don't warn about the shadowing pattern variable.

This means that there might be no type relationship between the two
symbols, and that elements of tuples might be swapped, i.e.,
there is no ordering relationship between arguments in the selector
and in the pattern.

For example, `(x, y) match { case (y, x) => }`.

For example, `t.toString match { case t => }`.

som-snytt commented:
My phrase for "shows up in scrutinee" is "implicated in selector expression".

I took your suggestion and scrapped the restrictions around the selector at the cost of more false negatives.

This is clearly wrong:
```
def f(x: Any) = x.toString match { case x => }
```
The previous enforcement was that the result of `x.f` must conform to `x` to intend "don't warn, I'm just refining x". I could still add that back.

I tried adding a type condition on the pat var and the shadowee, but the typical challenge is
```
def f(x: Option[X]) = x match { case Some(x) => }
```
Maybe it's enough to note type of x in the wrapped type. That is, like the current tweak, it's enough if it appears somewhere, and we won't be picky about the shape. Really we're saying, x should have some type relationship to what it shadows, if not conforming then at least embedding.

som-snytt commented:
@lrytz that was my last swing for now. You may not like that I chose not to clutter up the help text. Currently, the lint passes the "no-brainer" test (not noisy and helpful if it warns).

Since everyone asks about warnings on Scala 2 vs Scala 3, not to mention compiler options, what is needed is that doc, which would explain the options operationally.

lrytz reviewed (approved):
Great, thank you for going through the many iterations!
